docs/docs/render-props.md

```diff
@@ -0,0 +1,226 @@
+
+For example, the following component tracks the mouse cursor position in a web app:
+
+```js
+class MouseTracker extends React.Component {
+  constructor(props) {
+    super(props);
+    this.handleMouseMove = this.handleMouseMove.bind(this);
+    this.state = { x: 0, y: 0 };
+  }
+
+  handleMouseMove(event) {
+    this.setState({
+      x: event.clientX,
+      y: event.clientY
+    });
+  }
+
+  render() {
+    return (
+      <div style={{ height: '100%' }} onMouseMove={this.handleMouseMove}>
+        <h1>Move the mouse around!</h1>
+        <p>The current mouse position is ({this.state.x}, {this.state.y})</p>
+      </div>
+    );
+  }
+}
+```
+
+As the mouse cursor moves around the screen, the component displays its (x, y) coordinates in a `<p>`.
+
+Now the question is: How can we reuse this behavior in another component? In other words, if another component needs to know about the mouse position, can we encapsulate that behavior so that we can easily share it with that component?
+
+Since components are the basic unit of code reuse in React, let's try refactoring the code a bit to use a `<Mouse>` component that encapsulates the behavior we need to reuse elsewhere.
+
+```js
+// The <Mouse> component encapsulates the behavior we need...
+class Mouse extends React.Component {
+  constructor(props) {
+    super(props);
+    this.handleMouseMove = this.handleMouseMove.bind(this);
+    this.state = { x: 0, y: 0 };
+  }
+
+  handleMouseMove(event) {
+    this.setState({
+      x: event.clientX,
+      y: event.clientY
+    });
+  }
+
+  render() {
+    return (
+      <div style={{ height: '100%' }} onMouseMove={this.handleMouseMove}>
+
+        {/* ...but how do we render something other than a <p>? */}
+        <p>The current mouse position is ({this.state.x}, {this.state.y})</p>
+      </div>
+    );
+  }
+}
+
+class MouseTracker extends React.Component {
+  render() {
+    return (
+      <div>
+        <h1>Move the mouse around!</h1>
+        <Mouse/>
+      </div>
+    );
+  }
+}
+```
+
+Now the `<Mouse>` component encapsulates all behavior associated with tracking the `mousemove` events and storing the (x, y) position of the cursor, but it's not yet truly reusable.
+
+For example, let's say we have a `<Cat>` component that renders the image of a cat chasing the mouse cursor around the screen. We might use a `<Cat mouse={{ x, y }}>` prop to tell the component the coordinates of the mouse so it knows where to position the image on the screen.
+
+As a first pass, you might try rendering the `<Cat>` *inside `<Mouse>`'s `render` method*, like this:
+
+```js
+class Cat extends React.Component {
+  render() {
+    const mouse = this.props.mouse
+    return (
+      <img src="/cat.jpg" style={{ position: 'absolute', left: mouse.x, top: mouse.y }}/>
+    );
+  }
+}
+
+class MouseWithCat extends React.Component {
```

Comment:
This would be a really weird way to reuse code.

I'd expect someone who didn't understand how `render` functions work would use the component as a variable.

```typescript
<Mouse
  component = { Cat }
/>

---

render() {
  const {
    component: Component,
  } = this.props;

  const {
    mouse,
  } = this.state;

  return (
    <Component
      mouse = { mouse }
    />
  );
}
```

Of course, `render` functions are more flexible because they don't make assumptions about the shape of the child component.  But, if someone wasn't using a `render` function